File: packages/Konclave.ProtocolContracts.TypeScript/src/membership.ts
import { create } from '@bufbuild/protobuf';
import {
  bytesKey,
  decodeBounded,
  encodeBounded,
  MAX_APPLICATION_MESSAGE_BYTES,
  MAX_CONSUMED_INVITATIONS,
  MAX_MEMBERS,
  MAX_RELAY_PAYLOAD_BYTES,
  required,
  validateFixedBytes,
  validateLengthRange,
  validateRole,
  validateRepeatedFieldLimits,
  validateUint64,
  validateVersion,
} from './common.js';
import { protocolErrorCodes, ProtocolValidationError } from './error.js';
import { decodeJoinProof, encodeJoinProof } from './identity.js';
import { ConversationRole } from './generated/konclave/protocol/v1/common_pb.js';
import {
  ConversationStateSchema,
  MembershipCommitBundleSchema,
  MembershipChangeSchema,
  MembershipControlSchema,
  type ConversationState,
  type MembershipCommitBundle,
  type MembershipChange,
} from './generated/konclave/protocol/v1/membership_pb.js';
import type { JoinProof } from './generated/konclave/protocol/v1/identity_pb.js';

const stateContract = 'ConversationState';
const changeContract = 'MembershipChange';
const controlContract = 'MembershipControl';
const commitBundleContract = 'MembershipCommitBundle';

export interface DecodedMembershipControl {
  membershipChange: MembershipChange;
  joinProof?: JoinProof;
}

/**
 * Encodes validated application-authorized conversation state.
 *
 * @throws {ProtocolValidationError} When the state violates a v1 bound or invariant.
 */
export function encodeConversationState(value: ConversationState): Uint8Array {
  validateConversationState(value);
  return encodeBounded(
    ConversationStateSchema,
    value,
    MAX_APPLICATION_MESSAGE_BYTES,
    stateContract,
  );
}

/**
 * Decodes and validates application-authorized conversation state.
 *
 * @throws {ProtocolValidationError} When the bytes are malformed, oversized, or invalid.
 */
export function decodeConversationState(bytes: Uint8Array): ConversationState {
  validateRepeatedFieldLimits(bytes, MAX_APPLICATION_MESSAGE_BYTES, stateContract, [
    { field: 'members', fieldNumber: 4, maximum: MAX_MEMBERS },
    {
      field: 'consumed_invitation_ids',
      fieldNumber: 5,
      maximum: MAX_CONSUMED_INVITATIONS,
    },
  ]);
  const value = decodeBounded(
    ConversationStateSchema,
    bytes,
    MAX_APPLICATION_MESSAGE_BYTES,
    stateContract,
  );
  validateConversationState(value);
  return value;
}

/**
 * Encodes one validated application-authorized membership transition.
 *
 * @throws {ProtocolValidationError} When the transition violates a v1 bound or invariant.
 */
export function encodeMembershipChange(value: MembershipChange): Uint8Array {
  validateMembershipChange(value);
  return encodeBounded(
    MembershipChangeSchema,
    value,
    MAX_APPLICATION_MESSAGE_BYTES,
    changeContract,
  );
}

/**
 * Decodes one application-authorized membership transition.
 *
 * MLS sender authentication and administrator authorization remain the caller's
 * responsibility.
 *
 * @throws {ProtocolValidationError} When the bytes are malformed, oversized, or invalid.
 */
export function decodeMembershipChange(bytes: Uint8Array): MembershipChange {
  const value = decodeBounded(
    MembershipChangeSchema,
    bytes,
    MAX_APPLICATION_MESSAGE_BYTES,
    changeContract,
  );
  validateMembershipChange(value);
  return value;
}

/**
 * Encodes membership context before MLS application encryption.
 *
 * These bytes contain membership plaintext and must never enter a relay envelope
 * without MLS encryption.
 *
 * @throws {ProtocolValidationError} When a nested contract is invalid or oversized.
 */
export function encodeMembershipControl(
  membershipChange: MembershipChange,
  joinProof?: JoinProof,
): Uint8Array {
  const value = create(MembershipControlSchema, {
    membershipChange: encodeMembershipChange(membershipChange),
    joinProof: joinProof ? encodeJoinProof(joinProof) : new Uint8Array(),
  });
  return encodeBounded(
    MembershipControlSchema,
    value,
    MAX_APPLICATION_MESSAGE_BYTES,
    controlContract,
  );
}

/**
 * Decodes membership context after MLS application decryption.
 *
 * @throws {ProtocolValidationError} When bytes are malformed, missing, or invalid.
 */
export function decodeMembershipControl(bytes: Uint8Array): DecodedMembershipControl {
  const value = decodeBounded(
    MembershipControlSchema,
    bytes,
    MAX_APPLICATION_MESSAGE_BYTES,
    controlContract,
  );
  requireOpaqueBytes(
    value.membershipChange,
    MAX_APPLICATION_MESSAGE_BYTES,
    'membership_control.membership_change',
  );
  return {
    membershipChange: decodeMembershipChange(value.membershipChange),
    joinProof: value.joinProof.length > 0 ? decodeJoinProof(value.joinProof) : undefined,
  };
}

/**
 * Encodes opaque encrypted membership control and its bound MLS Commit.
 *
 * @throws {ProtocolValidationError} When either field is empty or the aggregate
 * relay payload is oversized.
 */
export function encodeMembershipCommitBundle(
  encryptedControl: Uint8Array,
  mlsCommit: Uint8Array,
): Uint8Array {
  requireOpaqueBytes(
    encryptedControl,
    MAX_RELAY_PAYLOAD_BYTES,
    'membership_commit_bundle.encrypted_control',
  );
  requireOpaqueBytes(mlsCommit, MAX_RELAY_PAYLOAD_BYTES, 'membership_commit_bundle.mls_commit');
  return encodeBounded(
    MembershipCommitBundleSchema,
    create(MembershipCommitBundleSchema, { encryptedControl, mlsCommit }),
    MAX_RELAY_PAYLOAD_BYTES,
    commitBundleContract,
  );
}

/**
 * Decodes client-only opaque framing from a relay envelope.
 *
 * @throws {ProtocolValidationError} When bytes are malformed, missing, or oversized.
 */
export function decodeMembershipCommitBundle(bytes: Uint8Array): MembershipCommitBundle {
  const value = decodeBounded(
    MembershipCommitBundleSchema,
    bytes,
    MAX_RELAY_PAYLOAD_BYTES,
    commitBundleContract,
  );
  requireOpaqueBytes(
    value.encryptedControl,
    MAX_RELAY_PAYLOAD_BYTES,
    'membership_commit_bundle.encrypted_control',
  );
  requireOpaqueBytes(
    value.mlsCommit,
    MAX_RELAY_PAYLOAD_BYTES,
    'membership_commit_bundle.mls_commit',
  );
  return value;
}

function requireOpaqueBytes(bytes: Uint8Array, maximum: number, field: string): void {
  if (bytes.length === 0) {
    throw new ProtocolValidationError(protocolErrorCodes.missingField, `${field} is missing`);
  }
  validateLengthRange(bytes.length, 1, maximum, field);
}

function validateConversationState(value: ConversationState): void {
  validateVersion(value.version, stateContract);
  validateFixedBytes(
    required(value.conversationId, 'conversation_id').value,
    32,
    'conversation_id',
  );
  validateUint64(value.epoch, 'epoch', false);
  validateLengthRange(value.members.length, 1, MAX_MEMBERS, 'members');
  validateLengthRange(
    value.consumedInvitationIds.length,
    0,
    MAX_CONSUMED_INVITATIONS,
    'consumed_invitation_ids',
  );

  const memberIds = new Set<string>();
  let hasAdministrator = false;
  for (const member of value.members) {
    const deviceId = required(member.deviceId, 'member.device_id').value;
    validateFixedBytes(deviceId, 32, 'member.device_id');
    validateRole(member.role);
    validateUint64(member.joinedEpoch, 'joined_epoch', false);
    if (member.joinedEpoch > value.epoch) {
      throw new ProtocolValidationError(
        protocolErrorCodes.memberJoinedAfterStateEpoch,
        'member joined after the represented conversation epoch',
      );
    }
    const key = bytesKey(deviceId);
    if (memberIds.has(key)) {
      throw new ProtocolValidationError(
        protocolErrorCodes.duplicateIdentifier,
        'member_device_id contains a duplicate identifier',
      );
    }
    memberIds.add(key);
    hasAdministrator ||= member.role === ConversationRole.ADMINISTRATOR;
  }
  if (!hasAdministrator) {
    throw new ProtocolValidationError(
      protocolErrorCodes.missingAdministrator,
      'conversation membership requires an administrator',
    );
  }

  const invitationIds = new Set<string>();
  for (const invitationId of value.consumedInvitationIds) {
    validateFixedBytes(invitationId.value, 16, 'consumed_invitation_id');
    const key = bytesKey(invitationId.value);
    if (invitationIds.has(key)) {
      throw new ProtocolValidationError(
        protocolErrorCodes.duplicateIdentifier,
        'consumed_invitation_id contains a duplicate identifier',
      );
    }
    invitationIds.add(key);
  }
}

function validateMembershipChange(value: MembershipChange): void {
  validateVersion(value.version, changeContract);
  validateFixedBytes(
    required(value.conversationId, 'conversation_id').value,
    32,
    'conversation_id',
  );
  validateUint64(value.parentEpoch, 'parent_epoch', false);
  validateFixedBytes(required(value.operationId, 'operation_id').value, 16, 'operation_id');

  switch (value.change.case) {
    case 'add':
      validateFixedBytes(
        required(value.change.value.deviceId, 'add.device_id').value,
        32,
        'add.device_id',
      );
      validateRole(value.change.value.role);
      validateFixedBytes(
        required(value.change.value.invitationId, 'add.invitation_id').value,
        16,
        'add.invitation_id',
      );
      validateFixedBytes(value.change.value.credentialBindingHash, 32, 'credential_binding_hash');
      return;
    case 'remove':
      validateFixedBytes(
        required(value.change.value.deviceId, 'remove.device_id').value,
        32,
        'remove.device_id',
      );
      return;
    case 'changeRole':
      validateFixedBytes(
        required(value.change.value.deviceId, 'change_role.device_id').value,
        32,
        'change_role.device_id',
      );
      validateRole(value.change.value.role);
      return;
    case undefined:
      throw new ProtocolValidationError(
        protocolErrorCodes.missingVariant,
        'membership_change.change is missing',
      );
  }
}

File: crates/Konclave.ProtocolContracts/src/v1/membership.rs
use KonclaveDomainCore::{
    AddMember, ChangeMemberRole, ConversationState, JoinProof, MAX_APPLICATION_MESSAGE_BYTES,
    MAX_CONSUMED_INVITATIONS, MAX_MEMBERS, MAX_RELAY_PAYLOAD_BYTES, Member,
    MembershipAuthorization, MembershipChange, RemoveMember,
};

use super::identity::{decode_join_proof, encode_join_proof};
use crate::KonclaveProtocolError;
use crate::v1::common::{
    conversation_id_from_wire, conversation_id_to_wire, credential_hash_from_bytes,
    credential_hash_to_bytes, decode_bounded, device_id_from_wire, device_id_to_wire,
    encode_bounded, invitation_id_from_wire, invitation_id_to_wire, operation_id_from_wire,
    operation_id_to_wire, require_repeated_field_limits, required, role_from_wire, role_to_wire,
    version_from_wire, version_to_wire,
};
use crate::wire::v1 as wire;

const STATE_CONTRACT: &str = "ConversationState";
const CHANGE_CONTRACT: &str = "MembershipChange";
const CONTROL_CONTRACT: &str = "MembershipControl";
const COMMIT_BUNDLE_CONTRACT: &str = "MembershipCommitBundle";

/// Opaque MLS application control plus its bound MLS Commit.
pub struct MembershipCommitBundleBytes {
    encrypted_control: Vec<u8>,
    mls_commit: Vec<u8>,
}

impl MembershipCommitBundleBytes {
    /// Returns the MLS application PrivateMessage containing membership context.
    #[must_use]
    pub fn encrypted_control(&self) -> &[u8] {
        &self.encrypted_control
    }

    /// Returns the MLS Commit PrivateMessage.
    #[must_use]
    pub fn mls_commit(&self) -> &[u8] {
        &self.mls_commit
    }
}

/// Encodes application-authorized conversation membership state.
///
/// # Errors
///
/// Returns a size error when the encoded state exceeds the v1 application limit.
pub fn encode_conversation_state(
    value: &ConversationState,
) -> Result<Vec<u8>, KonclaveProtocolError> {
    let wire = wire::ConversationState {
        version: Some(version_to_wire(value.version())),
        conversation_id: Some(conversation_id_to_wire(value.conversation_id())),
        epoch: value.epoch(),
        members: value
            .members()
            .iter()
            .map(|member| wire::Member {
                device_id: Some(device_id_to_wire(member.device_id())),
                role: role_to_wire(member.role()),
                joined_epoch: member.joined_epoch(),
            })
            .collect(),
        consumed_invitation_ids: value
            .consumed_invitation_ids()
            .iter()
            .copied()
            .map(invitation_id_to_wire)
            .collect(),
    };
    encode_bounded(&wire, MAX_APPLICATION_MESSAGE_BYTES, STATE_CONTRACT)
}

/// Decodes and validates conversation membership state.
///
/// # Errors
///
/// Returns a typed protocol or domain validation error.
pub fn decode_conversation_state(bytes: &[u8]) -> Result<ConversationState, KonclaveProtocolError> {
    require_repeated_field_limits(
        bytes,
        MAX_APPLICATION_MESSAGE_BYTES,
        STATE_CONTRACT,
        [
            (4, MAX_MEMBERS, "members"),
            (5, MAX_CONSUMED_INVITATIONS, "consumed_invitation_ids"),
        ],
    )?;
    let wire: wire::ConversationState =
        decode_bounded(bytes, MAX_APPLICATION_MESSAGE_BYTES, STATE_CONTRACT)?;
    require_collection_bound(wire.members.len(), 1, MAX_MEMBERS, "members")?;
    require_collection_bound(
        wire.consumed_invitation_ids.len(),
        0,
        MAX_CONSUMED_INVITATIONS,
        "consumed_invitation_ids",
    )?;
    let members = wire
        .members
        .into_iter()
        .map(|member| {
            Ok(Member::new(
                device_id_from_wire(member.device_id)?,
                role_from_wire(member.role)?,
                member.joined_epoch,
            ))
        })
        .collect::<Result<Vec<_>, KonclaveProtocolError>>()?;
    let consumed_invitation_ids = wire
        .consumed_invitation_ids
        .into_iter()
        .map(|id| invitation_id_from_wire(Some(id)))
        .collect::<Result<Vec<_>, _>>()?;
    Ok(ConversationState::new(
        version_from_wire(wire.version, STATE_CONTRACT)?,
        conversation_id_from_wire(wire.conversation_id)?,
        wire.epoch,
        members,
        consumed_invitation_ids,
    )?)
}

fn require_collection_bound(
    actual: usize,
    minimum: usize,
    maximum: usize,
    field: &'static str,
) -> Result<(), KonclaveProtocolError> {
    if actual < minimum || actual > maximum {
        return Err(KonclaveDomainCore::KonclaveDomainError::OutOfRange {
            field,
            minimum,
            maximum,
            actual,
        }
        .into());
    }
    Ok(())
}

/// Encodes one application-authorized membership transition.
///
/// # Errors
///
/// Returns a size error when the encoded transition exceeds the v1 application limit.
pub fn encode_membership_change(
    value: &MembershipAuthorization,
) -> Result<Vec<u8>, KonclaveProtocolError> {
    let change = match value.change() {
        MembershipChange::Add(add) => wire::membership_change::Change::Add(wire::AddMember {
            device_id: Some(device_id_to_wire(add.device_id())),
            role: role_to_wire(add.role()),
            invitation_id: Some(invitation_id_to_wire(add.invitation_id())),
            credential_binding_hash: credential_hash_to_bytes(add.credential_binding_hash()),
        }),
        MembershipChange::Remove(remove) => {
            wire::membership_change::Change::Remove(wire::RemoveMember {
                device_id: Some(device_id_to_wire(remove.device_id())),
            })
        }
        MembershipChange::ChangeRole(change) => {
            wire::membership_change::Change::ChangeRole(wire::ChangeMemberRole {
                device_id: Some(device_id_to_wire(change.device_id())),
                role: role_to_wire(change.role()),
            })
        }
    };
    let wire = wire::MembershipChange {
        version: Some(version_to_wire(value.version())),
        conversation_id: Some(conversation_id_to_wire(value.conversation_id())),
        parent_epoch: value.parent_epoch(),
        operation_id: Some(operation_id_to_wire(value.operation_id())),
        change: Some(change),
    };
    encode_bounded(&wire, MAX_APPLICATION_MESSAGE_BYTES, CHANGE_CONTRACT)
}

/// Decodes and validates one application-authorized membership transition.
///
/// Cryptographic sender authentication and administrator authorization remain the
/// caller's responsibility.
///
/// # Errors
///
/// Returns a typed protocol or domain validation error.
pub fn decode_membership_change(
    bytes: &[u8],
) -> Result<MembershipAuthorization, KonclaveProtocolError> {
    let wire: wire::MembershipChange =
        decode_bounded(bytes, MAX_APPLICATION_MESSAGE_BYTES, CHANGE_CONTRACT)?;
    let change = match required(wire.change, "membership_change.change")? {
        wire::membership_change::Change::Add(add) => MembershipChange::Add(AddMember::new(
            device_id_from_wire(add.device_id)?,
            role_from_wire(add.role)?,
            invitation_id_from_wire(add.invitation_id)?,
            credential_hash_from_bytes(&add.credential_binding_hash)?,
        )),
        wire::membership_change::Change::Remove(remove) => {
            MembershipChange::Remove(RemoveMember::new(device_id_from_wire(remove.device_id)?))
        }
        wire::membership_change::Change::ChangeRole(change) => {
            MembershipChange::ChangeRole(ChangeMemberRole::new(
                device_id_from_wire(change.device_id)?,
                role_from_wire(change.role)?,
            ))
        }
    };
    Ok(MembershipAuthorization::new(
        version_from_wire(wire.version, CHANGE_CONTRACT)?,
        conversation_id_from_wire(wire.conversation_id)?,
        wire.parent_epoch,
        operation_id_from_wire(wire.operation_id)?,
        change,
    ))
}

/// Encodes membership authorization and optional add-member proof for MLS encryption.
///
/// These bytes contain membership plaintext and must be encrypted as an MLS
/// application PrivateMessage before entering a relay envelope.
///
/// # Errors
///
/// Returns a protocol, domain, or size error.
pub fn encode_membership_control(
    authorization: &MembershipAuthorization,
    join_proof: Option<&JoinProof>,
) -> Result<Vec<u8>, KonclaveProtocolError> {
    let membership_change = encode_membership_change(authorization)?;
    let join_proof = join_proof.map(encode_join_proof).transpose()?;
    let wire = wire::MembershipControl {
        membership_change: membership_change.into(),
        join_proof: join_proof.unwrap_or_default().into(),
    };
    encode_bounded(&wire, MAX_APPLICATION_MESSAGE_BYTES, CONTROL_CONTRACT)
}

/// Decodes membership context after MLS application decryption.
///
/// Cryptographic sender authentication and transition authorization remain the
/// caller's responsibility.
///
/// # Errors
///
/// Returns a malformed, missing-field, protocol, domain, or size error.
pub fn decode_membership_control(
    bytes: &[u8],
) -> Result<(MembershipAuthorization, Option<JoinProof>), KonclaveProtocolError> {
    let wire: wire::MembershipControl =
        decode_bounded(bytes, MAX_APPLICATION_MESSAGE_BYTES, CONTROL_CONTRACT)?;
    if wire.membership_change.is_empty() {
        return Err(KonclaveProtocolError::MissingField {
            field: "membership_control.membership_change",
        });
    }
    let authorization = decode_membership_change(&wire.membership_change)?;
    let join_proof = (!wire.join_proof.is_empty())
        .then(|| decode_join_proof(&wire.join_proof))
        .transpose()?;
    Ok((authorization, join_proof))
}

/// Encodes an opaque encrypted control message and MLS Commit for relay transport.
///
/// # Errors
///
/// Returns a missing-field or aggregate relay-payload size error.
pub fn encode_membership_commit_bundle(
    encrypted_control: &[u8],
    mls_commit: &[u8],
) -> Result<Vec<u8>, KonclaveProtocolError> {
    require_non_empty(
        encrypted_control,
        "membership_commit_bundle.encrypted_control",
    )?;
    require_non_empty(mls_commit, "membership_commit_bundle.mls_commit")?;
    let wire = wire::MembershipCommitBundle {
        encrypted_control: encrypted_control.to_vec().into(),
        mls_commit: mls_commit.to_vec().into(),
    };
    encode_bounded(&wire, MAX_RELAY_PAYLOAD_BYTES, COMMIT_BUNDLE_CONTRACT)
}

/// Decodes the opaque client-only framing carried by a relay envelope.
///
/// The returned MLS messages remain cryptographically untrusted until processed by
/// `CryptographicCore`.
///
/// # Errors
///
/// Returns a malformed, missing-field, or aggregate relay-payload size error.
pub fn decode_membership_commit_bundle(
    bytes: &[u8],
) -> Result<MembershipCommitBundleBytes, KonclaveProtocolError> {
    let wire: wire::MembershipCommitBundle =
        decode_bounded(bytes, MAX_RELAY_PAYLOAD_BYTES, COMMIT_BUNDLE_CONTRACT)?;
    require_non_empty(
        &wire.encrypted_control,
        "membership_commit_bundle.encrypted_control",
    )?;
    require_non_empty(&wire.mls_commit, "membership_commit_bundle.mls_commit")?;
    Ok(MembershipCommitBundleBytes {
        encrypted_control: wire.encrypted_control.to_vec(),
        mls_commit: wire.mls_commit.to_vec(),
    })
}

fn require_non_empty(bytes: &[u8], field: &'static str) -> Result<(), KonclaveProtocolError> {
    if bytes.is_empty() {
        Err(KonclaveProtocolError::MissingField { field })
    } else {
        Ok(())
    }
}
